pkg/scheduler/controller/scheduler_control.go

```diff
@@ -0,0 +1,320 @@
+	// Determine all candidate seed clusters matching the shoot's cloud and region.
+	for _, seed := range seedList {
+		if seed.DeletionTimestamp == nil && seed.Spec.Cloud.Profile == shoot.Spec.Cloud.Profile && seed.Spec.Cloud.Region == shoot.Spec.Cloud.Region && seed.Spec.Visible != nil && *seed.Spec.Visible && verifySeedAvailability(seed) {
+			candidates = append(candidates, seed)
+		}
+	}
+	return candidates
+}
+
+func determineCandidatesWithMinimalDistanceStrategy(seeds []*gardenv1beta1.Seed, shoot *gardenv1beta1.Shoot, candidates []*gardenv1beta1.Seed) []*gardenv1beta1.Seed {
+	if candidates = determineCandidatesWithSameRegionStrategy(seeds, shoot, candidates); candidates != nil {
+		return candidates
+	}
+
+	var (
+		currentMaxMatchingCharacters int
+		shootRegion                  = shoot.Spec.Cloud.Region
+	)
+
+	// Determine all candidate seed clusters with matching cloud provider but different region that are lexicographically closest to the shoot
+	for _, seed := range seeds {
+		if seed.DeletionTimestamp == nil && seed.Spec.Cloud.Profile == shoot.Spec.Cloud.Profile && seed.Spec.Visible != nil && *seed.Spec.Visible && verifySeedAvailability(seed) {
+			seedRegion := seed.Spec.Cloud.Region
+
+			for currentMaxMatchingCharacters < len(shootRegion) {
+				if strings.HasPrefix(seedRegion, shootRegion[:currentMaxMatchingCharacters+1]) {
+					candidates = []*gardenv1beta1.Seed{}
+					currentMaxMatchingCharacters++
+					continue
+				} else if strings.HasPrefix(seedRegion, shootRegion[:currentMaxMatchingCharacters]) {
+					candidates = append(candidates, seed)
+				}
+				break
+			}
+		}
+	}
+	return candidates
+}
+
+func generateSeedUsageMap(shootList []*gardenv1beta1.Shoot) map[string]int {
+	m := map[string]int{}
+
+	for _, shoot := range shootList {
+		if seed := shoot.Spec.Cloud.Seed; seed != nil {
+			m[*seed]++
+		}
+	}
+
+	return m
+}
+
+func validateDisjointedNetworks(seed *gardenv1beta1.Seed, shoot *gardenv1beta1.Shoot) bool {
+	k8sNetworks, err := gardenhelper.GetK8SNetworks(shoot)
+	if err != nil || k8sNetworks == nil {
+		return false
+	}
+
+	allErrs := schedulerutils.ValidateNetworkDisjointedness(seed.Spec.Networks, *k8sNetworks, field.NewPath(""))
+	return len(allErrs) == 0
+}
+
+func verifySeedAvailability(seed *gardenv1beta1.Seed) bool {
+	if cond := gardencorev1alpha1helper.GetCondition(seed.Status.Conditions, gardenv1beta1.SeedAvailable); cond != nil {
+		return cond.Status == gardencorev1alpha1.ConditionTrue
+	}
+	return false
+}
+
+// UpdateShootToBeScheduledOntoSeed (ctx context.Context, cl client.Client, client kubernetes.Interface, shoot *gardenv1beta1.Shoot, seed *gardenv1beta1.Seed) error {
+func UpdateShootToBeScheduledOntoSeed(ctx context.Context, shoot *gardenv1beta1.Shoot, seed *gardenv1beta1.Seed, executeSchedulingRequest executeSchedulingRequest) error {
```

Comment:
This function is somehow strange. It just sets the seed name for the shoot and executes some arbitrary function that is passed in. Can you get rid of this / refactor this? Also, the documentation for this seems to be a bit off.

Reply:
Thanks. I adjusted the documenting comment.
Yes currently it only executes the actual update function that it gets passed in. 
Reason being that the caller can decide how it wants to update the shoot.

1) Scheduler Control: Use kutil.TryUpdateShoot() because retries are needed in case there was an external change to the resource -> happens because the ControllerManager is acting upon the Shoot at the same time (adding condition that it cannot reconcile atm).

2) Unit test:  use Mock client

3) Integration uses the garden client directly (however it could also use retry)